graylog2-server/src/test/java/org/graylog2/shared/buffers/processors/MessageTimestampTest.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.shared.buffers.processors;
+
+import org.graylog2.plugin.Message;
+import org.graylog2.plugin.MessageFactory;
+import org.graylog2.plugin.TestMessageFactory;
+import org.graylog2.plugin.Tools;
+import org.joda.time.DateTime;
+import org.junit.jupiter.api.Test;
+
+import java.time.Duration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class MessageTimestampTest {
+    private final MessageFactory messageFactory = new TestMessageFactory();
+
+    @Test
+    void testValid() {
+        DateTime now = Tools.nowUTC();
+        Message msg = messageFactory.createMessage("test message", "localhost", now);
+        normalize(msg, Duration.ofDays(30));
+        assertThat(msg.getTimestamp()).isEqualTo(now);
+    }
+
+    @Test
+    void testWithinGracePeriod() {
+        DateTime time = Tools.nowUTC().plusDays(10);
+        Message msg = messageFactory.createMessage("test message", "localhost", time);
+        normalize(msg, Duration.ofDays(30));
+        assertThat(msg.getTimestamp()).isEqualTo(time);
+    }
+
+    @Test
+    void testExceedsGracePeriod() {
+        DateTime now = Tools.nowUTC();
+        DateTime time = now.plusDays(50);
+        Message msg = messageFactory.createMessage("test message", "localhost", time);
+        normalize(msg, Duration.ofDays(30));
+        assertThat(fuzzyEquals(msg.getTimestamp(), now)).isTrue();
+    }
+
+    void normalize(Message msg, Duration gracePeriod) {
+        msg.ensureValidTimestamp();
+        msg.normalizeTimestamp(gracePeriod);
+    }
+
+    final static long fuzzyMillis = 100;
+
+    boolean fuzzyEquals(DateTime d1, DateTime d2) {
```

Comment:
We should never rely on timing in tests, especially when we control the entire code.
Instead please use a Clock (or something like https://github.com/Graylog2/graylog2-server/blob/f35df42e165ac570b8b27de3f8eeac85e74ed610/graylog2-server/src/main/java/org/graylog2/plugin/InstantMillisProvider.java#L25) and control the passage of time explicitly.
This makes the tests more robust and easier to reason about.

Reply:
fixed